File: Ethernet/c/mccEthernet.c
/*
 *  Copyright (c) 2015 Warren J. Jasper <wjasper@ncsu.edu>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include <stdio.h>
#include <unistd.h>
#include <stdbool.h>
#include <string.h>
#include <sys/select.h>
#include <sys/time.h>
#include <sys/types.h>
#include "mccEthernet.h"

static int recvfromTimeOut(int sock, struct timeval* timeout)
{
  fd_set fds;

  FD_ZERO(&fds);
  FD_SET(sock, &fds);
  // -1: error occurred
  // 0: timed out
  // >0: data ready to be read
  return select(sock+1, &fds, 0, 0, timeout);
}

void printDeviceInfo(EthernetDeviceInfo *device_info)
{
  printf("   Found device: %s\n", device_info->NetBIOS_Name);
  printf("   MAC: %02X:%02X:%02X:%02X:%02X:%02X\n",
	 device_info->MAC[0], device_info->MAC[1], device_info->MAC[2],
	 device_info->MAC[3], device_info->MAC[4], device_info->MAC[5]);
  printf("   IP: %s\n", inet_ntoa(device_info->Address.sin_addr));
  printf("   Product ID: 0x%04X\n", device_info->ProductID);
  printf("   FW version: %02X.%02X\n", 
	 (unsigned char)(device_info->FirmwareVersion >> 8),
	 (unsigned char)(device_info->FirmwareVersion));
  printf("   Boot version: %02X.%02X\n",
	 (unsigned char)(device_info->BootloaderVersion >> 8),
	 (unsigned char)(device_info->BootloaderVersion));
  printf("   Command Port: %d\n", device_info->CommandPort);
  printf("   Status: %s\n", (device_info->Status == 0) ? "Available" : "In Use");
  printf("   Remote host: %s\n\n", inet_ntoa(device_info->RemoteHost.sin_addr));
}

int discoverDevice(EthernetDeviceInfo *device_info, uint16_t productID)
{
  EthernetDeviceInfo device;  // a MCC device that responded
  struct sockaddr_in sendaddr;
  struct sockaddr_in recvaddr;
  struct sockaddr_in remoteaddr;
  struct timeval tv;
  
  int sock;
  int broadcast;
  unsigned char msg[64];
  bool finished = false;   
  int nfound = 0;            // number of matched devices found
  socklen_t remoteaddrSize;
  int BytesReceived;

  // create the socket
  if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
    perror("discoverDevice: Error in creating socket");
    return -1;
  }
  // set the broadcast option
  broadcast = 1;
  if (setsockopt(sock, SOL_SOCKET, SO_BROADCAST, &broadcast, sizeof(broadcast)) < 0) {
    perror("discoverDevice: Error setting socket options");
    close(sock);
    return -1;
  }
  // set up the send and receive addresses
  memset(&sendaddr, 0, sizeof(sendaddr));
  sendaddr.sin_family = AF_INET;
  sendaddr.sin_addr.s_addr = INADDR_BROADCAST;
  sendaddr.sin_port = htons(DISCOVER_PORT);
  
  memset(&recvaddr, 0, sizeof(recvaddr));
  recvaddr.sin_family = AF_INET;
  recvaddr.sin_addr.s_addr = INADDR_ANY;
  recvaddr.sin_port = htons(DISCOVER_PORT);

  // bind the socket for receive
  if (bind(sock, (struct sockaddr*)&recvaddr, sizeof(recvaddr)) < 0) {
    perror("discoverDevice: Error binding port");
    close(sock);
    return -1;
  }

  // send a broadcast discover datagram
  msg[0] = 'D';
  if (sendto(sock, msg, 1, 0, (struct sockaddr*)&sendaddr, sizeof(sendaddr)) != 1) {
    perror("discoverDevice: sendto failed");
    close(sock);
    return -1;
  }

  tv.tv_sec = 1;
  tv.tv_usec = 0;

  // look for replies (including the original broadcast)
  while (!finished) {
    switch (recvfromTimeOut(sock, &tv)) {
      case 0:
        // timed out
        finished = true;
        break;
      case -1:
        // error
        printf("Error from recvfromTimeOut\n");
        finished = true;
        close(sock);
        return -1;
        break;
      default:
       // got a reply
        remoteaddrSize = sizeof(remoteaddr);
        BytesReceived = recvfrom(sock, msg, 64, 0, (struct sockaddr*)&remoteaddr, &remoteaddrSize);
        if ((BytesReceived == 64) && (msg[0] == 'D')) {
	  memcpy(device.MAC, &msg[1], 6);
	  memcpy(&device.ProductID, &msg[7], 2);
	  memcpy(&device.FirmwareVersion, &msg[9], 2);
	  memcpy(device.NetBIOS_Name, &msg[11], 16);
	  memcpy(&device.CommandPort, &msg[27], 2);
	  memcpy(&device.Status, &msg[33], 2);
	  memcpy(&device.RemoteHost.sin_addr, &msg[35], 4);
	  memcpy(&device.BootloaderVersion, &msg[39], 2);
	  memcpy(&device.Address, &remoteaddr, sizeof(remoteaddr));
          if (device.ProductID == productID) {  // check for match
	    memcpy(device_info, &device, sizeof(EthernetDeviceInfo));
	    nfound++;
	  }
	}
	break;
    }
  }
  close(sock);
  return nfound;
}

int discoverDevices(EthernetDeviceInfo *devices_info[], uint16_t productID, int maxDevices)
{
  EthernetDeviceInfo device;  // a MCC device that responded
  struct sockaddr_in sendaddr;
  struct sockaddr_in recvaddr;
  struct sockaddr_in remoteaddr;
  struct timeval tv;
  int sock;
  int broadcast;
  unsigned char msg[64];
  bool finished = false;   
  int nfound = 0;            // number of matched devices found
  socklen_t remoteaddrSize;
  int BytesReceived;

  // create the socket
  if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
    perror("discoverDevices: Error in creating socket");
    return -1;
  }

  // set the broadcast option
  broadcast = 1;
  if (setsockopt(sock, SOL_SOCKET, SO_BROADCAST, &broadcast, sizeof(broadcast)) < 0) {
    perror("discoverDevices: Error setting socket options");
    close(sock);
    return -1;
  }

  // set up the send and receive addresses
  memset(&sendaddr, 0, sizeof(sendaddr));
  sendaddr.sin_family = AF_INET;
  sendaddr.sin_addr.s_addr = INADDR_BROADCAST;
  sendaddr.sin_port = htons(DISCOVER_PORT);
  
  memset(&recvaddr, 0, sizeof(recvaddr));
  recvaddr.sin_family = AF_INET;
  recvaddr.sin_addr.s_addr = INADDR_ANY;
  recvaddr.sin_port = htons(DISCOVER_PORT);

  // bind the socket for receive
  if (bind(sock, (struct sockaddr*)&recvaddr, sizeof(recvaddr)) < 0) {
    perror("discoverDevices: Error binding port");
    close(sock);
    return -1;
  }

  // send a broadcast discover datagram
  msg[0] = 'D';
  if (sendto(sock, msg, 1, 0, (struct sockaddr*)&sendaddr, sizeof(sendaddr)) != 1) {
    perror("discoverDevices: sendto failed");
    close(sock);
    return -1;
  }

  tv.tv_sec = 1;
  tv.tv_usec = 0;

  // look for replies (including the original broadcast)
  while (!finished) {
    switch (recvfromTimeOut(sock, &tv)) {
      case 0:
        // timed out
        finished = true;
        break;
      case -1:
        // error
        printf("Error from recvfromTimeOut\n");
        finished = true;
        close(sock);
        return -1;
        break;
      default:
       // got a reply
        remoteaddrSize = sizeof(remoteaddr);
        BytesReceived = recvfrom(sock, msg, 64, 0, (struct sockaddr*)&remoteaddr, &remoteaddrSize);
        if ((BytesReceived == 64) && (msg[0] == 'D')) {
	  memcpy(device.MAC, &msg[1], 6);
	  memcpy(&device.ProductID, &msg[7], 2);
	  memcpy(&device.FirmwareVersion, &msg[9], 2);
	  memcpy(device.NetBIOS_Name, &msg[11], 16);
	  memcpy(&device.CommandPort, &msg[27], 2);
	  memcpy(&device.Status, &msg[33], 2);
	  memcpy(&device.RemoteHost.sin_addr, &msg[35], 4);
	  memcpy(&device.BootloaderVersion, &msg[39], 2);
	  memcpy(&device.Address, &remoteaddr, sizeof(remoteaddr));
          if (device.ProductID == productID) {  // check for match
	    memcpy(devices_info[nfound], &device, sizeof(EthernetDeviceInfo));
	    devices_info[nfound]->connectCode = 0x0;  // default connect code
	    devices_info[nfound]->frameID = 0x0;      // initialize frameID
	    nfound++;
	    if (nfound > maxDevices) {
	      close(sock);
	      return nfound;
	    }
	  }
	}
	break;
    }
  }
  close(sock);
  return nfound;
}

int openDevice(uint32_t addr, uint32_t connectCode)
{
  int sock;
  struct sockaddr_in sendaddr;
  struct timeval tv;
  unsigned char msg[64];
  int bytesReceived;

  // open the UDP socket
  if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
    perror("openDevice: Error creating socket.");
      return -1;
  }

  // set up the send address
  memset(&sendaddr, 0, sizeof(sendaddr));
  sendaddr.sin_family = AF_INET;
  sendaddr.sin_addr.s_addr = addr;
  sendaddr.sin_port = htons(DISCOVER_PORT);

  if (connect(sock, (const struct sockaddr*) &sendaddr, sizeof(sendaddr)) < 0) {
    perror("openDevice: error in connect.");
    close(sock);
    return -1;
  }
 
  // send the connect message
  msg[0] = 'C';
  memcpy(&msg[1], &connectCode, 4);
  if (send(sock, msg, 5, 0) < 0) {
    perror("openDevice: Error in send.");
    close(sock);
    return -1;
  }

  tv.tv_sec = 1;
  tv.tv_usec = 0;

  // look for a reply
  switch (recvfromTimeOut(sock, &tv)) {
    case 0:  // timed out
    case -1: //  error    
      close (sock);
      return -1;
      break;
    default:  // got a reply
      bytesReceived = recv(sock, msg, 64, 0);
      if ((bytesReceived == 2) && (msg[0] = 'C') && (msg[1] == 0)) {
	break;
      } else {
	close(sock);
	return -1;
      }
      break;
  }
  close(sock);   // finished with the UDP portion

  if ((sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
    perror("openDevice: error creating TCP socket.");
    return -1;
  }

  // set up the send address
  memset(&sendaddr, 0, sizeof(sendaddr));
  sendaddr.sin_family = AF_INET;
  sendaddr.sin_addr.s_addr = addr;
  sendaddr.sin_port = htons(COMMAND_PORT);

  // create a tcp connection
  if ((connect(sock, (const struct sockaddr*) &sendaddr, sizeof(sendaddr))) < 0) {
      perror("openDevice: can not connect to device.");
      close(sock);
      return -1;
  }
    return sock;
}     

int flushInput(int sock)
{
  int numRecv, numTotal=0;
  char cbuf[512];

  while (1) {
      numRecv = recv(sock, cbuf, sizeof cbuf, MSG_DONTWAIT);
      if (numRecv <= 0) break;
      numTotal += numRecv;
  }
  if (numTotal > 0) fprintf(stderr, "ethernet::flushInput flushed %d bytes\n", numTotal);
  return numTotal;
}

int sendMessage(int sock, void *message, int length, int flags)
{
  flushInput(sock);
  return send(sock, message, length, flags);
}

int receiveMessage(int sock, void *message, int maxLength, unsigned long timeout)
{
  struct timeval tv;
  int bytesReceived = 0;

  if (sock < 0) {  // invalid socket number.
    return -1;
  }

  // set a receive timeout
  tv.tv_sec = timeout/1000;
  tv.tv_usec = (timeout - (tv.tv_sec*1000)) * 1000;

  // set a receive timeout  
  setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, (const char *) &tv, sizeof(tv));

  switch (recvfromTimeOut(sock, &tv)) {
    case 0:   // timed out
    case -1:  // error
      return -1;
      break;
    default:  // got a reply
      bytesReceived = recv(sock, message, maxLength, 0);
      break;
  }
  return bytesReceived;
}

unsigned char calcChecksum(void *buffer, int length)
{
  int i;
  unsigned char checksum = 0;

  for (i = 0; i < length; i++) {
    checksum += ((unsigned char*) buffer)[i];
  }
  return checksum;
}

File: Ethernet/python/mccEthernet.py
#! /usr/bin/python3
#
# Copyright (c) 2018 Warren J. Jasper <wjasper@ncsu.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import socket
import sys

# Global constants
DISCOVER_PORT =   54211
COMMAND_PORT =    54211
SCAN_PORT =       54212
    
MSG_SUCCESS =         0 # Command succeeded
MSG_ERROR_PROTOCOL =  1 # Command failed due to improper protocol (number of expected data bytes did not match protocol definition)
MSG_ERROR_PARAMETER = 2 # Command failed due to invalid parameters (the data contents were incorrect)
MSG_ERROR_BUSY =      3 # Command failed because the resource was busy
MSG_ERROR_READY =     4 # Command failed because the resource was not ready
MSG_ERROR_TIMEOUT =   5 # Command failed due to a resource timeout
MSG_ERROR_OTHER =     6 # Command failed due to some other error

MSG_HEADER_SIZE =     6
MSG_CHECKSUM_SIZE =   1

MSG_INDEX_START =      0
MSG_INDEX_COMMAND =    1
MSG_INDEX_FRAME =      2
MSG_INDEX_STATUS =     3
MSG_INDEX_COUNT_LOW =  4  # The maximum value for count is 1024
MSG_INDEX_COUNT_HIGH = 5
MSG_INDEX_DATA =       6

MSG_REPLY =            0x80
MSG_START =            0xDB

# Base class for lookup tables of calibration coefficients (slope and offset)
class table:
  def __init__(self):
    self.slope = 0.0
    self.intercept = 0.0

# Error Handling
class Error(Exception):
  ''' Base class for other exceptions.'''
  pass

class ResultError(Error):
  ''' Raised when return TCP packet fails'''
  pass

# Definitions for basic MCC Ethernet Device class
class mccEthernetDevice:

  def __init__(self, productID=None, device_address=None):
    self.commandPort = 54211
    if (device_address != None):
      self.address = device_address
    else:
      self.address = ''
    self.frameID = 0
    self.MAC = 0x0
    self.productID = productID
    self.NetBIOS = ''
    self.firmwareVersion = 0x0
    self.bootloadVersion = 0x0
    self.connectCode = 0x0
    return

  def printDeviceInfo(self):
    print('  Device Name: ', self.NetBIOS)
    print('  IP address:', self.address)
    print('  Product ID:', hex(self.productID))
    print('  Command Port:', self.commandPort)
    print('  MAC:', hex(self.MAC>>40&0xff)[2:].zfill(2)+':'+\
          hex(self.MAC>>32&0xff)[2:]+':'+\
          hex(self.MAC>>24&0xff)[2:]+':'+\
          hex(self.MAC>>16&0xff)[2:]+':'+\
         hex(self.MAC>>8&0xff)[2:]+':'+\
          hex(self.MAC&0xff)[2:])
    print('  Boot version:', str(self.bootloadVersion>>8)+'.'+str(self.bootloadVersion&0xff))
    print('  Firmware version:', str(self.firmwareVersion>>8)+'.'+str(self.firmwareVersion&0xff))
    print('')

  def mccOpenDevice(self, connectCode=0):
    # open the UDP socket
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)     # create the socket
    try:
      s.connect((self.address, DISCOVER_PORT))
    except:
      raise
    # send the connect message
    msg = bytearray(5)
    msg[0] = ord('C')
    msg[1] = (connectCode>>24) & 0xff
    msg[2] = (connectCode>>16) & 0xff
    msg[3] = (connectCode>>8)  & 0xff
    msg[4] = connectCode       & 0xff

    s.settimeout(1)          # set noblocking 1 second timeout
    s.send(msg)
    try:
      data = s.recv(10)
    except socket.timeout:
      raise
      print('mccOpenDevice: timeout error.\n')
      s.close()
      return
    s.close()
    if (len(data) == 2 and chr(data[0]) == 'C' and bytes(data)[1] == 0x0):
      # connection made
      pass
    elif bytes(data)[1] == 0x1:
      print('mccOpenDevice: Incorrect connect code.\n')
      return
    elif bytes(data)[1] == 0x2:
      print('mccOpenDevice: conntion ignored.\n')
      return
    elif bytes(data)[1] == 0x3:
      print('mccOpenDevice: device in use.\n')
      return
    else:
      print('mccOpenDevice: unknown value.\n')
      return
    
    # finished with the UDP portion.

    # create connection with TCP socket
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
      s.connect((self.address, COMMAND_PORT))
      self.sock = s
      return
    except:
      print('Error connecting to Ethernet Device')
      return

  def calcChecksum(self, buf, length):
    checksum = 0
    for i in range(length):
      checksum += buf[i]
    return (checksum & 0xff)

  def flushInput(self):
    # Flush input buffers.
    numTotal = 0
    while True:
      try:
        cbuf = self.sock.recv(512, self.sock.MSG_DONTWAIT)
        numTotal += len(cbuf)
      except:
        break
    if numTotal > 0:
      print('flushInput flushed', numTotal, 'bytes')
    return numTotal

  def sendMessage(self, message, flush=True):
    if (flush):
      self.flushInput()
    self.sock.send(message)
  

###############  end class defition ##########################

def mccDiscover(productID=None):
  devices = []
  nfound = 0
  s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)     # create the socket
  s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)  # set the broadcast option
  s.settimeout(1.0)                                        # set noblocking 1 second timeout
  s.bind(('', DISCOVER_PORT))                              # bind the socket for receive
  msg = 'D'
  #send a broadcast discover datagram
  s.sendto(msg.encode(), ('<broadcast>', DISCOVER_PORT))
  while True:
    try:
      msg, address = s.recvfrom(1024)
      if (len(msg) == 64):
        id = (msg[7] | (msg[8]<<8))
        if (id == productID or productID == None):
          devices.append(mccEthernetDevice())
          devices[nfound].address = address[0]
          devices[nfound].MAC = 0x0
          for i in range(1,7):
            devices[nfound].MAC += (msg[i]<<((6-i)*8))
          devices[nfound].NetBIOS = msg[11:26].decode()
          devices[nfound].productID = msg[7] | (msg[8]<<8)
          devices[nfound].firmwareVersion = (msg[9] | (msg[10]<<8))
          devices[nfound].bootloadVersion = (msg[39] | ( msg[40]<<8))
          devices[nfound].frameID = 0
          devices[nfound].ConnectCode = 0x0
          nfound += 1
    except socket.timeout:
      s.close()
      return devices


# Structures for Temperature 
##################################################
# NIST Thermocouple coefficients
#
# The following types are supported:
#
#    J, K, T, E, R, S, B, N
#
# Define the types of Thermocouples supported */

CHAN_DISABLE  = 0
TC_TYPE_J     = 1   # Type J thermocouple
TC_TYPE_K     = 2   # Type K thermocouple
TC_TYPE_T     = 3   # Type T thermocouple
TC_TYPE_E     = 4   # Type E thermocouple
TC_TYPE_R     = 5   # Type R thermocouple
TC_TYPE_S     = 6   # Type S thermocouple
TC_TYPE_B     = 7   # Type B thermocouple
TC_TYPE_N     = 8   # Type N thermocouple
